### ISSUE TITLE
Consider of there might be a way to execute generated code on the server

### ISSUE THREAD

**Issue by TimLethbridge** [umple/umple#1622]
See if there is a way to enable students to write models with a main program that they could execute directly on the server. It shouldn't be done on the main Umple compilation server, however, but by a companion server.

It would be necessary to both sandbox any such execution (to prevent security breaches) and also time-limit it so an infinite loop wouldn't cause problems. 

Sandboxing could perhaps be accomplished by a technology such as this:
https://www.usenix.org/system/files/conference/atc13/atc13-kim.pdf
It would only work on Linux, but that would be OK.

Time-limiting could be done by creating a pool of JVMs and sending code for execution. Any JVM that exceeds time limits could be killed and replaced. The ulimit function could be used for this. Also code that has got out of control could be prevented from running again (based on a comparison).

Another strategy might be to generate Javascript and run the code in the user's browser.

This is low priority, but has been requested by several users.

**Comment by TimLethbridge** [comment]
See issue #1634 (run Javac at least). That would improve usability and would be a first step towards this

**Comment by TimLethbridge** [comment]
This issue is a duplicate of #1394. I am going to close #1394

**Comment by TimLethbridge** [comment]
Another suggestion is to fire up Docker images that could run generated code, but would be resource limited and sandboxed. If one gets out of control it could be terminated. Kubernetes could be used to orchestrate these.

**Comment by love1024** [comment]
There are different ways using which we can implement this functionality. I have given an overview of different approaches that I am considering:
__Virtual Machine:__  
This would be too slow and not possible with Umple, so not going to look into this.  
__Using Linux:__  
Various online judges such as Codeforces, Codechef, Topcoder, etc are running on a large scale and providing functionality to run a user's code in a sandbox environment. They are securing the sandbox environment using Linux and Java restrictions such as giving limited memory, restricting network access, using chroot, etc.  
However, their source code is not open-sourced and I am looking into some open-source implementations to get a better understanding such as DOMJudge (https://github.com/DOMjudge/domjudge).   
__Docker Containers:__  
We can use it independently or on top of Linux restrictions to secure the environment more. Securing execution of Java code in a container requires certain policies to be applied on a Docker Container such as AppArmor policies (https://docs.docker.com/engine/security/apparmor/)  

In conclusion, This is just an overview of the initial idea and I am exploring the second and third approaches and will keep posting my analysis on this issue.

**Comment by TimLethbridge** [comment]
The plan is enable UmpleOnline to communicate with a Docker image that would execute Java code that has been generated already successfully.

This feature will work in UmpleOnline. So the developer working on this need to understand the flow in UmpleOnline (line numbers indicated below are as of January 2022).

I suggest that the mechanism to compile Java remain unchanged. If execution is to happen, then the compile mechanism would be used, and if successful, then a separate Ajax call would be made to execute it.

We will start analysis by looking in umpleonline/scripts:

Calls are initiated from the front-end: umple_action.js calls Action.ajax (that is at line 2771). 
   - which in turn is called from Action.generateCode (line 1293)
   - (later, when results come back, Action.generateCodeCallback is called)
   - We will likely have to create an enhanced Action.executeCode that will only be called if generating has happened.

At the back end, the work triggered by the above starts with compiler.php
  - The relevant block of code for generating code starts at line 324, which ends up calling executeCommand with an UmpleSync.jar command line at line 355.
  - We would need to add another block to trigger compiling the Java and executing (always assuming that generated code is present).
  
executeCommmand is in compiler_config.php at line 782
  
Originally executeCommand, was very straightforwad. But a few years ago we make an important change: If the command starts with 'java -jar umplesync.jar' (see line 792), an attempt is made to send the arguments to a running umplesync.jar server through a socket, instead of starting a new JVM. This is critical for performance. The serverRun function does this, which is at line 820.

Whether the call to generate code is done through a call to a running server, or just as command line arguments, the results would be the same ... if there are no errors, the code is generated.

UmpleSync.jar does NOT currently have the needed functionality to tell you which main program to execute. UmpleSync.jar would need to be enhanced to output the main program. Let us turn our attention to directory cruise.umple/src

The two main methods are in Main_Code.ump. The one used for umple.jar (which is used on the command line) will output the main method found on the command line. This happens at line 376 after a call to CodeCompiler.getMainClasses(model).

It will be necessary to get umplesync.jar to somehow call this same method and store the result so it can be used by the executing enhancement. See below.

Umplesync.jar uses PlaygroundMain starting at line 713.

When generating code, it sets up the needed variables and then calls model.run() at line 1117
If generation was successful, the code is obtained from the model in line 1168, although it is also already saved on the disk. One strategy (Strategy 1) would be to encode the main method in the result from model.getCode() so this can be processed by compiler.php

So back in compiler.php the output is in $sourceCode (as of line 367). It is output to the screen around line 550. In strategy 1, the encoded main method could first be extracted, and placed somewhere where the execute command would be able to access it.

Strategy 2 would be for the model.run to write out a file with the main methods when compiling Java. In this case a variable would have to be set in the model to ensure that the 'run' method, before exiting, calls getMainClasses and outputs the result to a file, perhaps called "JavaMainClasses.txt". The executing mechanism could then just read this. I suggest that this would be the best approach.

So for early prototypes. just create a file manually called  "JavaMainClasses.txt" and have the execution method read this.
